ruby/ql/lib/codeql/ruby/controlflow/BasicBlocks.qll

```diff
@@ -4,52 +4,49 @@ private import codeql.ruby.AST
 private import CfgNodes
 private import SuccessorTypes
+private import CfgImpl::BasicBlocks as BasicBlocksImpl
 
 /**
  * A basic block, that is, a maximal straight-line sequence of control flow nodes
  * without branches or joins.
  */
-class BasicBlock extends TBasicBlockStart {
-  /** Gets the scope of this basic block. */
-  final CfgScope getScope() { result = this.getFirstNode().getScope() }
-
+final class BasicBlock extends BasicBlocksImpl::BasicBlock {
   /** Gets an immediate successor of this basic block, if any. */
-  BasicBlock getASuccessor() { result = this.getASuccessor(_) }
+  BasicBlock getASuccessor() { result = super.getASuccessor() }
 
   /** Gets an immediate successor of this basic block of a given type, if any. */
-  BasicBlock getASuccessor(SuccessorType t) {
-    result.getFirstNode() = this.getLastNode().getASuccessor(t)
-  }
+  BasicBlock getASuccessor(SuccessorType t) { result = super.getASuccessor(t) }
 
   /** Gets an immediate predecessor of this basic block, if any. */
-  BasicBlock getAPredecessor() { result.getASuccessor() = this }
+  BasicBlock getAPredecessor() { result = super.getAPredecessor() }
 
   /** Gets an immediate predecessor of this basic block of a given type, if any. */
-  BasicBlock getAPredecessor(SuccessorType t) { result.getASuccessor(t) = this }
+  BasicBlock getAPredecessor(SuccessorType t) { result = super.getAPredecessor(t) }
 
-  /** Gets the control flow node at a specific (zero-indexed) position in this basic block. */
-  CfgNode getNode(int pos) { bbIndex(this.getFirstNode(), result, pos) }
+  // The overrides below are to use `CfgNode` instead of `CfgImpl::Node`
+  CfgNode getNode(int pos) { result = super.getNode(pos) }
 
-  /** Gets a control flow node in this basic block. */
-  CfgNode getANode() { result = this.getNode(_) }
+  CfgNode getANode() { result = super.getANode() }
 
   /** Gets the first control flow node in this basic block. */
-  CfgNode getFirstNode() { this = TBasicBlockStart(result) }
+  CfgNode getFirstNode() { result = super.getFirstNode() }
 
   /** Gets the last control flow node in this basic block. */
-  CfgNode getLastNode() { result = this.getNode(this.length() - 1) }
-
-  /** Gets the length of this basic block. */
-  int length() { result = strictcount(this.getANode()) }
+  CfgNode getLastNode() { result = super.getLastNode() }
 
   /**
    * Holds if this basic block immediately dominates basic block `bb`.
    *
-   * That is, all paths reaching basic block `bb` from some entry point
-   * basic block must go through this basic block (which is an immediate
-   * predecessor of `bb`).
+   * That is, this basic block is the unique basic block satisfying:
+   * 1. This basic block strictly dominates `bb`
+   * 2. There exists no other basic block that is strictly dominated by this
+   *    basic block and which strictly dominates `bb`.
+   *
+   * All basic blocks, except entry basic blocks, have a unique immediate
+   * dominator.
```

Comment:
I've updated a bunch of the existing qdocs related to immediate dominance. I think this is more precise/correct.

The previous explanations stated that the immediately dominating node would be an immediate predecessor, which I don't think is correct. See for instance [this example](https://en.wikipedia.org/wiki/Dominator_(graph_theory)#/media/File:Dominator_control_flow_graph.svg) from Wikipedia where 2 immediately dominates 5, but is not a direct predecessor. Evaluating the predicates also gives immediate dominators that are not direct predecessors.